### PR TITLE
Add Laravel 6.0 compatibility.

### LINKED ISSUE #834: Add Laravel 6.0 compatibility.
Adding this issue as a placeholder until the package can be updated. I will attempt to submit a PR if I get the time.

### PR DESCRIPTION
**Pull request by mikebronner**
Fixes #834

### CODE REVIEW

**Comment by ArlonAntonius** [comment]
PR without any valid changes, closing.

**Comment by mikebronner** [comment]
@ArlonAntonius I am working on the changes (notice the WIP icon in the title). Adding them soon. Please reopen, as it won't take my changes unless the PR is open.

File: tests/unit-tests/Filesystem/LoadsTranslationsTest.php
Comment (mikebronner) [inline tests/unit-tests/Filesystem/LoadsTranslationsTest.php:97]: Not sure if this will impact the package? I removed this check, because the `password` entry no longer exists in Laravel 6.

**Comment by mikebronner** [comment]
@ArlonAntonius Not sure why the CircleCI tests have a failure. Does that have something to do with the setup of CircleCI? I'm not getting these errors when running locally. Anything you recommend I should do? Thanks!

**Comment by mikebronner** [comment]
@ArlonAntonius other than the CirceCI issue, I think this is ready for review.

**Comment by ArlonAntonius** [comment]
Thanks for taking the time @mikebronner, took a quick look at the PR and a few things stand out that I want to have a closer look at later next week.

The error that this PR is getting is related to Laravel/Helpers not being installed (as array_get is a helper function). If you're able to solve that (I personally would prefer without installing laravel/helpers) then you should get green tests.

**Comment by mikebronner** [comment]
Will look into that in the morning :) Found where it is in the code.

**Comment by mikebronner** [comment]
@ArlonAntonius I made the fix for `array_get`, but a new sql error popped up. I'm not sure how to debug this, as I don't run apache locally. I tested with both mysql and pgsql, and received no errors. Any suggestions what might be causing this?

**Comment by thijskok** [comment]
> @ArlonAntonius I made the fix for `array_get`, but a new sql error popped up. I'm not sure how to debug this, as I don't run apache locally. I tested with both mysql and pgsql, and received no errors. Any suggestions what might be causing this?

Hi @mikebronner, forgive the intrusion, but I might have some additional information (looking forward to a 6.0 compatible version 👍).

The CircleCI tests fail because of a failed password update attempt ([right here](https://github.com/tenancy/multi-tenant/blob/963987bd90266b3f7d2d0e9af35a97d0d346f8be/src/Generators/Webserver/Database/Drivers/MariaDB.php#L115)). In MySQL 5.7.6 and MariaDB 10.1.20, the way passwords should be changed was... well, changed :) . This kinda complicates things, because updating password now becomes MySQL / MariaDB version dependent.

Found [a tutorial illustrating the problem at DigitalOcean](https://www.digitalocean.com/community/tutorials/how-to-reset-your-mysql-or-mariadb-root-password#step-4-%E2%80%94-changing-the-root-password).

**Comment by mikebronner** [comment]
@thijskok Thanks for the information! As this issue seems to go a bit beyond this PR, would it be OK if I left fixing the password issue to you, @ArlonAntonius? This seems to be more related to the testing setup and such. It appears that most or all PRs are now failing with this error.

File: src/Database/Console/Migrations/FreshCommand.php
Comment (mikebronner) [inline src/Database/Console/Migrations/FreshCommand.php]: I added this work-around, because the `dropAllTables()` method no longer exists in Laravel 6. Feel free to optimize. :)

Comment (drbyte) [reply src/Database/Console/Migrations/FreshCommand.php]: Looks about how I'd do it, too. 👍 

Ref to Laravel 6 commit: https://github.com/laravel/framework/commit/c26b435d826ff7882a0185eb5e48001683d48b9d
Er, the PR was https://github.com/laravel/framework/pull/29620

**Comment by ArlonAntonius** [comment]
@mikebronner 
Does the array fix actually work? I remember Laravel getting rid of all helper functions and putting them in a separate package that was not required by laravel/laravel by default.

Uhmmm regarding the password changing, I'm not sure what MySQL version etc our CircleCI is running now, but we never meant to update it and suddenly it's broken and I think it's cause we're on a version we're not expected to be on. I'm not sure what approach will be.

Would rather leave the password changing up to @luceos or @fletch3555 who might have a better vision on this issue in this package than I do.

**Comment by mikebronner** [comment]
@ArlonAntonius Not sure what you mean? I'm no longer relying on the helpers and am using the recommended `Arr` package provided by Laravel. Could you elaborate? As recommended by the upgrade guide: https://laravel.com/docs/6.0/upgrade#helpers

All the `Str` and `Arr` functions remain in Laravel, just the helper methods that were used to access them are removed. So I opted to access that one method directly, instead of pulling in the external helpers package.

**Comment by ArlonAntonius** [comment]
Ahh yeah oke seems fine, my bad.

**Comment by drbyte** [comment]
New MariaDB 10.4 password approach: passwords no longer required since they want you to connect via Unix Socket. They're highly discouraging the use of setting actual passwords.
Docs: https://mariadb.org/authentication-in-mariadb-10-4/

A bug report has already been filed for tenancy/multi-tenant: #826

**Comment by drbyte** [comment]
FWIW the Laravel 6 changes appear to be fine. 

It's unfortunate that the CircleCI config just happens to be grabbing an incompatible MariaDB instance. Probably still worth forcing it to an older container inside this PR to get it to pass though.

**Comment by thijskok** [comment]
> @thijskok Thanks for the information! As this issue seems to go a bit beyond this PR, would it be OK if I left fixing the password issue to you, @ArlonAntonius? This seems to be more related to the testing setup and such. It appears that most or all PRs are now failing with this error.

You’re welcome! This issue does affect every PR created recently. To provide some context: 5.7.6 was released in March 2015, MariaDB 10.1.20 in December 2016.

A separate PR fixing this issue would be “cleaner”, but also delays this PR 🤔

**Comment by mikebronner** [comment]
But isn't this a bigger issue with CircleCI configuration? Wouldn't that have to be updated to not use a password on the database, or is everything I need in the package code? If so, I'll be happy to try to address this. (I don't run MariaDB, so not sure if I'm the best person to implement this fix, but I can try.)

**Comment by drbyte** [comment]
3 things happening here:

1. Laravel 6 updates: fine
2. CircleCI started using a newer default image for MariaDB: 10.4.
3. This package needs MariaDB updates for compatibility with MariaDB 10.4 -- See #826 

Fixing '3' should fix '2' automatically, and allow the tests for '1' to pass.
Or patching '2' by downgrading CircleCI's MariaDB container image should allow '1' to pass, and allow the others to be addressed externally in another PR.

File: tests/unit-tests/Filesystem/LoadsTranslationsTest.php
Comment (luceos) [reply tests/unit-tests/Filesystem/LoadsTranslationsTest.php:97]: Don't remove this, you just invalidated part of our tests. Pick any other translation key and match it up please.

File: composer.json
Comment (luceos) [inline composer.json]: We do not offer backward compatibility. Simply move forward to 6.x please.

File: src/Database/Console/Migrations/FreshCommand.php
Comment (luceos) [reply src/Database/Console/Migrations/FreshCommand.php]: Don't add backward compatibility logic. Just remove that and use db:wipe. Potentially remove fresh command completely and implement a tenancy:db:wipe one?

**Review by luceos — CHANGES_REQUESTED** [review]
Thanks for all the hard work, I've not been in the loop about all changes related to 6.x and as such extremely thankful for this 👏

**Comment by drbyte** [comment]
The last passing CircleCI tests were for `mariadb:10.3.16-bionic`

A quick temporary fix for this PR to run CircleCI tests on 10.3 is to edit the yml file and change the 2 references for `image: mariadb:latest` on lines 50 and 54 
to `image: mariadb:10.3.17-bionic`

File: tests/unit-tests/Filesystem/LoadsTranslationsTest.php
Comment (mikebronner) [reply tests/unit-tests/Filesystem/LoadsTranslationsTest.php:97]: @luceos  Can you elaborate? Won't picking a random translation key break the test? What is the goal of this check -- perhaps by better understanding what is being tested for, I can implement a proper solution.

Would you like me to test any other of the items, like so:
```php
$this->assertEquals($original['reset'], trans('passwords.reset', [], 'en'));
```

**Comment by luceos** [comment]
We should most definitely use the last images. Not sure what changed, would gladly merge a pr that fixes this.

**Comment by mikebronner** [comment]
@drbyte That I can do! :) (scratch that ... x-nay on the downgrade-ay)

@luceos As this is a wider issue not directly related to this PR, I would feel more comfortable if that is handled outside of this PR, as we are creeping on scope. This issue currently affects all PRs, I would feel much more comfortable if that could be something you (or another contributor that is using MariaDB) could address.

I'll work on the other things you mentioned, and not downgrade the MariaDB version in this PR for now.

**Comment by drbyte** [comment]
IMO since tenancy/multi-tenant doesn't yet properly support MariaDB 10.4 at the time of this PR, the CircleCI dependency in this PR SHOULD be downgraded (since it's a hard fail in this package) so this PR can be merged.

It makes no sense to hold back Laravel 6.0 compatibility due to a CI infrastructure dependency incompatibility in unrelated code.
See #826 

Then, separately, someone can rewrite the [MariaDB code](https://github.com/tenancy/multi-tenant/blob/963987bd90266b3f7d2d0e9af35a97d0d346f8be/src/Generators/Webserver/Database/Drivers/MariaDB.php#L115) to support newer MariaDB capabilities, and then @luceos can decide how/when to tag that new compatibility capability.

**Comment by fletch3555** [comment]
Agreed on the version change.  Let's lock MariaDB at 10.3 for now.

**Comment by mikebronner** [comment]
@luceos @ArlonAntonius Ready for re-review. I have not done anything with the MariaDB issue, waiting for your feedback on my concerns and the suggestions from others above.

File: phpunit.xml
Comment (drbyte) [inline phpunit.xml]: Were these phpunit.xml changes intentional?

File: composer.json
Comment (bkintanar) [inline composer.json]: @mikebronner Thank you for this PR. Do we have a package that needs dev stability?

Comment (luceos) [reply composer.json]: Great find @bkintanar this should be removed now that the dev dependency to Laravel was removed 👍

**Comment by luceos** [comment]
I am okay with locking mariadb for now, I'd love to see an upgrade of the mariadb driver in this package as soon as possible though.

@mikebronner

**Comment by mikebronner** [comment]
Hi guys, I'm going to need someone else to tackle the MariaDB issue and take it from here, I just don't have the resources to take of that at the moment. I believe everything else is now in order.

**Comment by drbyte** [comment]
I did a PR for locking the MariaDB stuff for CI: #838

**Comment by fletch3555** [comment]
No worries @mikebronner.  We'll track that elsewhere

**Comment by fletch3555** [comment]
@mikebronner I merged #838.  Can you update your branch?

**Comment by mikebronner** [comment]
@fletch3555 certainly, coming right up.

**Comment by ArlonAntonius** [comment]
Thanks for all the work on this PR so far everybody, real team effort 😄 
I still feel like there are too many dependencies that are wildcards at this moment, can we revert those to the previous case as much as possible?

**Comment by mikebronner** [comment]
@ArlonAntonius I'll take a look, brb.

File: composer.json
Comment (fletch3555) [inline composer.json]: Can we revert the alphabetizing of these? It's hard to tell what the actual version change was when the entire line appears changed.

File: tests/unit-tests/Commands/RunCommandTest.php
Comment (fletch3555) [inline tests/unit-tests/Commands/RunCommandTest.php]: Why the change from Exception to Error?

**Review by fletch3555 — CHANGES_REQUESTED** [review]

**Comment by drbyte** [comment]
@ArlonAntonius @fletch3555 @luceos Can you tag a 5.4.5 release before merging and tagging this laravel 6 stuff?  There are important fixes already merged that haven't been released. And this PR forces a major version requirement that shouldn't impinge on those other fixes.

**Comment by fletch3555** [comment]
@drbyte I don't see any reason not to.  Thanks for the reminder!

File: tests/unit-tests/Commands/RunCommandTest.php
Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: Laravel 6 throws an error, instead of an exception, here. Feel free to run the test with it set to exception, perhaps I misinterpreted it, and the error is due to something else?

File: composer.json
Comment (mikebronner) [reply composer.json]: Roger.

**Comment by drbyte** [comment]
... and if it fits your plans, maybe this Laravel 6 PR can be part of a newer 5.5 series release, just to keep compatibility issues separate?

File: tests/unit-tests/Commands/RunCommandTest.php
Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: Here's the output:
```
1) Hyn\Tenancy\Tests\Commands\RunCommandTest::proxies_exceptions
Failed asserting that exception of type "Error" matches expected exception "Exception". Message was: "Call to undefined method Exception::run()" at
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Console/Command.php:231
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Console/Command.php:204
/Users/mike/Developer/Sites/multi-tenant/src/Commands/RunCommand.php:82
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php:39
/Users/mike/Developer/Sites/multi-tenant/src/Commands/RunCommand.php:85
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:32
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:90
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:34
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Container/Container.php:591
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Console/Command.php:192
/Users/mike/Developer/Sites/multi-tenant/vendor/symfony/console/Command/Command.php:255
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Console/Command.php:179
/Users/mike/Developer/Sites/multi-tenant/vendor/symfony/console/Application.php:915
/Users/mike/Developer/Sites/multi-tenant/vendor/symfony/console/Application.php:272
/Users/mike/Developer/Sites/multi-tenant/vendor/symfony/console/Application.php:148
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Console/Application.php:90
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Console/Application.php:182
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:273
/Users/mike/Developer/Sites/multi-tenant/vendor/laravel/framework/src/Illuminate/Foundation/Testing/Concerns/InteractsWithConsole.php:43
/Users/mike/Developer/Sites/multi-tenant/tests/unit-tests/Commands/RunCommandTest.php:64
```

**Comment by drbyte** [comment]
@mikebronner your phpunit.xml snuck back in again ... :P

File: tests/unit-tests/Commands/RunCommandTest.php
Comment (fletch3555) [reply tests/unit-tests/Commands/RunCommandTest.php]: That looks like it's throwing an `Error` because you're calling a method that doesn't exist on that class.  In this case, it looks like `$this->resolveCommand()` is returning an `Exception` instead of a `Command` (https://github.com/laravel/framework/blob/6.x/src/Illuminate/Console/Command.php#L231).  At quick glance, it seems the only way that can happen is if the object passed in is already an `Exception`

Comment (fletch3555) [reply tests/unit-tests/Commands/RunCommandTest.php]: Can you attempt to log the contents of `$this->argument('run')` right before that call? https://github.com/tenancy/multi-tenant/blob/5.x/src/Commands/RunCommand.php#L81

If it looks good, step in and repeat.  You'll probably have to try to reproduce the failure locally first, assuming you haven't already

Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: Output of `dump($this->argument('run'), $options->toArray());` is:
```
"exception"
[]
```

Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: I'm not sure what you are expecting back. Is `exception` a valid method? I would assume not? If not, the response seems appropriate to me.

Comment (drbyte) [reply tests/unit-tests/Commands/RunCommandTest.php]: In theory, here in the test, calling `run` with `run = exception` is trying to execute the `exception` command defined in RunCommandTest on lines 30-32-ish.
... so I'd expect it to throw `\Exception` ... hmmm

Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: `dump($this->resolveCommand($command));` is what returns the exception:
```
Exception {#1354
  #message: ""
  #code: 0
  #file: "./vendor/laravel/framework/src/Illuminate/Console/Command.php"
  #line: 248
  trace: {
    ./vendor/laravel/framework/src/Illuminate/Console/Command.php:248 { …}
    ./vendor/laravel/framework/src/Illuminate/Console/Command.php:230 { …}
    ./vendor/laravel/framework/src/Illuminate/Console/Command.php:204 { …}
    ./src/Commands/RunCommand.php:82 {
      ›     $this->argument('run'),
      ›     $options->toArray()
      › );
      arguments: {
        $command: "exception"
        $arguments: []
      }
    }
    ./vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php:39 { …}
    ./src/Commands/RunCommand.php:85 { …}
    Hyn\Tenancy\Commands\RunCommand->handle() {}
    ./vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:32 { …}
    ./vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:90 { …}
    ./vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:34 { …}
    ./vendor/laravel/framework/src/Illuminate/Container/Container.php:591 { …}
    ./vendor/laravel/framework/src/Illuminate/Console/Command.php:192 { …}
    ./vendor/symfony/console/Command/Command.php:255 { …}
    ./vendor/laravel/framework/src/Illuminate/Console/Command.php:179 { …}
    ./vendor/symfony/console/Application.php:915 { …}
    ./vendor/symfony/console/Application.php:272 { …}
    ./vendor/symfony/console/Application.php:148 { …}
    ./vendor/laravel/framework/src/Illuminate/Console/Application.php:90 { …}
    ./vendor/laravel/framework/src/Illuminate/Console/Application.php:182 { …}
    ./vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:273 { …}
    ./vendor/laravel/framework/src/Illuminate/Foundation/Testing/Concerns/InteractsWithConsole.php:43 { …}
    ./tests/unit-tests/Commands/RunCommandTest.php:63 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestCase.php:1327 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestCase.php:947 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestResult.php:691 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestCase.php:679 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestSuite.php:568 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestSuite.php:568 { …}
    ./vendor/phpunit/phpunit/src/Framework/TestSuite.php:568 { …}
    ./vendor/phpunit/phpunit/src/TextUI/TestRunner.php:616 { …}
    ./vendor/phpunit/phpunit/src/TextUI/Command.php:201 { …}
    ./vendor/phpunit/phpunit/src/TextUI/Command.php:160 { …}
    ./vendor/phpunit/phpunit/phpunit:61 { …}
  }
}
```

Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: Thanks @drbyte that provided the info I needed. Implementing fix now.

Comment (fletch3555) [reply tests/unit-tests/Commands/RunCommandTest.php]: The issue is that it's passing in "exception", which [this code](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Console/Command.php#L248) then tries to instantiate.  So you get an `\Exception` object which isn't a `Command` and therefore doesn't have a `->run(...)` method

Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: Hmm, I thought I had something, but it was a red herring. I'm coming up empty. The reason an error is being thrown is because it doesn't recognize the dynamically created command. It doesn't appear to exist.

Comment (fletch3555) [reply tests/unit-tests/Commands/RunCommandTest.php]: so the issue is that our test is running with `exception` as an "invalid" command name.  We should update the test to something like `commandThatDoesNotExist`

Comment (mikebronner) [reply tests/unit-tests/Commands/RunCommandTest.php]: @fletch3555 Bingo!

Comment (fletch3555) [reply tests/unit-tests/Commands/RunCommandTest.php]: My guess is that it worked before because the code was different: https://github.com/laravel/framework/blob/5.8/src/Illuminate/Console/Command.php#L213

Instead of instantiating the class directly, it tries to pull it from IOC, which I'm guessing threw an exception instead of blindly continuing

Comment (drbyte) [reply tests/unit-tests/Commands/RunCommandTest.php]: Excellent!

**Comment by mikebronner** [comment]
Please let me know if there is anything else I can help with. Thanks @drbyte for the help, especially the MariaDB aspect.

**Comment by luceos** [comment]
great job @mikebronner @drbyte @ArlonAntonius @fletch3555 @bkintanar 

We'll be moving to 5.5.x with this.

**Pull request merged**
